File: reimplementation/adl2/crates/adl-syntax/src/dump.rs
//! Canonical `--dump-ast` text form: stable, deterministic, reviewable
//! (SPEC_ARCHITECTURE §3). Snapshot tests lock this format.

use crate::ast::*;
use crate::span::LineMap;
use std::fmt::Write as _;

/// Render the canonical AST dump. `src` is needed to express spans as
/// line:col (byte offsets would churn on whitespace-only edits).
#[must_use]
pub fn dump_ast(src: &str, file: &File) -> String {
    let map = LineMap::new(src);
    let mut d = Dumper {
        map,
        out: String::new(),
        depth: 0,
    };
    d.line("File");
    d.depth += 1;
    for section in &file.sections {
        d.section(section);
    }
    d.out
}

struct Dumper {
    map: LineMap,
    out: String,
    depth: usize,
}

impl Dumper {
    fn line(&mut self, text: &str) {
        for _ in 0..self.depth {
            self.out.push_str("  ");
        }
        self.out.push_str(text);
        self.out.push('\n');
    }

    fn at(&self, span: crate::span::Span) -> String {
        let (l, c) = self.map.line_col(span.start);
        format!("@{l}:{c}")
    }

    fn nested(&mut self, header: &str, f: impl FnOnce(&mut Self)) {
        self.line(header);
        self.depth += 1;
        f(self);
        self.depth -= 1;
    }

    fn section(&mut self, s: &Section) {
        match s {
            Section::Info(info) => {
                let header = format!("Info name={} {}", info.name.name, self.at(info.span));
                self.nested(&header, |d| {
                    for line in &info.lines {
                        let items: Vec<String> = line.items.iter().map(info_item).collect();
                        d.line(&format!(
                            "Line key={} items=[{}]",
                            line.key.name,
                            items.join(", ")
                        ));
                    }
                });
            }
            Section::Table(t) => {
                let header = format!(
                    "Table name={} type={} nvars={} errors={} {}",
                    t.name.name,
                    t.table_type.name,
                    t.nvars,
                    t.errors,
                    self.at(t.span)
                );
                self.nested(&header, |d| {
                    let vals: Vec<String> = t.values.iter().map(NumLit::canon).collect();
                    d.line(&format!("Values [{}]", vals.join(" ")));
                });
            }
            Section::CountsFormat(cf) => {
                let header = format!("CountsFormat name={} {}", cf.name.name, self.at(cf.span));
                self.nested(&header, |d| {
                    for p in &cf.processes {
                        let cols: Vec<&str> = p.columns.iter().map(|c| c.name.as_str()).collect();
                        d.line(&format!(
                            "Process name={} title={:?} columns=[{}]",
                            p.name.name,
                            p.title.value,
                            cols.join(", ")
                        ));
                    }
                });
            }
            Section::Define(def) => {
                let header = format!(
                    "Define kw={} name={} {}",
                    def.keyword,
                    def.name.name,
                    self.at(def.span)
                );
                self.nested(&header, |d| d.expr(&def.body));
            }
            Section::Object(obj) => {
                let header = format!(
                    "Object kw={} name={} {}",
                    obj.keyword.as_str(),
                    obj.name.name,
                    self.at(obj.span)
                );
                self.nested(&header, |d| {
                    for stmt in &obj.stmts {
                        d.object_stmt(stmt);
                    }
                });
            }
            Section::Region(region) => {
                let header = format!(
                    "Region kw={} name={} {}",
                    region.keyword.as_str(),
                    region.name.name,
                    self.at(region.span)
                );
                self.nested(&header, |d| {
                    for stmt in &region.stmts {
                        d.region_stmt(stmt);
                    }
                });
            }
        }
    }

    fn object_stmt(&mut self, stmt: &ObjectStmt) {
        match stmt {
            ObjectStmt::Take {
                keyword,
                source,
                binders,
                alias,
                span,
            } => {
                let mut header = format!("Take kw={keyword}");
                match source {
                    TakeSource::Ident(id) => {
                        let _ = write!(header, " src={}", id.name);
                    }
                    TakeSource::Call { name, .. } => {
                        let _ = write!(header, " src=call:{}", name.name);
                    }
                    TakeSource::Union { members, .. } => {
                        let names: Vec<&str> = members.iter().map(|m| m.name.as_str()).collect();
                        let _ = write!(header, " src=union({})", names.join(","));
                    }
                }
                if !binders.is_empty() {
                    let names: Vec<&str> = binders.iter().map(|b| b.name.as_str()).collect();
                    let _ = write!(header, " binders=[{}]", names.join(","));
                }
                if let Some(a) = alias {
                    let _ = write!(header, " alias={}", a.name);
                }
                let _ = write!(header, " {}", self.at(*span));
                if let TakeSource::Call { args, .. } = source {
                    let header = header.clone();
                    self.nested(&header, |d| {
                        for arg in args {
                            d.arg(arg);
                        }
                    });
                } else {
                    self.line(&header);
                }
            }
            ObjectStmt::Cut {
                keyword,
                cond,
                span,
            } => {
                let header = format!("Cut kw={keyword} {}", self.at(*span));
                self.nested(&header, |d| d.expr(cond));
            }
            ObjectStmt::Reject { cond, span } => {
                let header = format!("Reject {}", self.at(*span));
                self.nested(&header, |d| d.expr(cond));
            }
        }
    }

    fn region_stmt(&mut self, stmt: &RegionStmt) {
        match stmt {
            RegionStmt::Cut {
                keyword,
                cond,
                span,
            } => {
                let header = format!("Cut kw={keyword} {}", self.at(*span));
                self.nested(&header, |d| d.expr(cond));
            }
            RegionStmt::Reject { cond, span } => {
                let header = format!("Reject {}", self.at(*span));
                self.nested(&header, |d| d.expr(cond));
            }
            RegionStmt::RegionRef(id) => {
                let header = format!("RegionRef name={} {}", id.name, self.at(id.span));
                self.line(&header);
            }
            RegionStmt::Bin { label, body, span } => {
                let mut header = "Bin".to_string();
                if let Some(l) = label {
                    let _ = write!(header, " label={:?}", l.value);
                }
                let _ = write!(header, " {}", self.at(*span));
                self.nested(&header, |d| match body {
                    BinBody::Boundaries { var, edges } => {
                        let es: Vec<String> = edges.iter().map(NumLit::canon).collect();
                        d.nested(&format!("Boundaries edges=[{}]", es.join(" ")), |d| {
                            d.expr(var);
                        });
                    }
                    BinBody::Cond(cond) => d.expr(cond),
                });
            }
            RegionStmt::Trigger { cond, span } => {
                let header = format!("Trigger {}", self.at(*span));
                self.nested(&header, |d| d.expr(cond));
            }
            RegionStmt::Histo {
                name,
                title,
                args,
                span,
            } => {
                let header = format!(
                    "Histo name={} title={:?} {}",
                    name.name,
                    title.value,
                    self.at(*span)
                );
                self.nested(&header, |d| {
                    for arg in args {
                        match arg {
                            HistoArg::Num(n) => d.line(&format!("Num {}", n.canon())),
                            HistoArg::NumList(ns) => {
                                let es: Vec<String> = ns.iter().map(NumLit::canon).collect();
                                d.line(&format!("NumList [{}]", es.join(" ")));
                            }
                            HistoArg::Expr(e) => d.expr(e),
                        }
                    }
                });
            }
            RegionStmt::Weight { name, value, span } => {
                let header = format!("Weight name={} {}", name.name, self.at(*span));
                self.nested(&header, |d| match value {
                    WeightValue::Num(n) => d.line(&format!("Num {}", n.canon())),
                    WeightValue::Expr(e) => d.expr(e),
                });
            }
            RegionStmt::Save {
                name,
                format,
                args,
                span,
            } => {
                let header = format!(
                    "Save name={} format={} {}",
                    name.name,
                    format.name,
                    self.at(*span)
                );
                self.nested(&header, |d| {
                    for arg in args {
                        d.arg(arg);
                    }
                });
            }
            RegionStmt::Print { args, span } => {
                let header = format!("Print {}", self.at(*span));
                self.nested(&header, |d| {
                    for arg in args {
                        d.arg(arg);
                    }
                });
            }
            RegionStmt::Counts {
                format,
                items,
                span,
            } => {
                self.line(&format!(
                    "Counts format={} items=[{}] {}",
                    format.name,
                    items.join(" "),
                    self.at(*span)
                ));
            }
            RegionStmt::Sort { raw, span } => {
                self.line(&format!(
                    "Sort (unsupported) raw={raw:?} {}",
                    self.at(*span)
                ));
            }
            RegionStmt::TypeTag { value, span } => {
                self.line(&format!("TypeTag value={} {}", value.name, self.at(*span)));
            }
        }
    }

    fn arg(&mut self, arg: &Arg) {
        match arg {
            Arg::Expr(e) => self.expr(e),
            Arg::Str(s) => self.line(&format!("Str {:?}", s.value)),
            Arg::Path(p) => self.line(&format!("Path {:?}", p.value)),
        }
    }

    fn expr(&mut self, e: &Expr) {
        match e {
            Expr::Num(n) => self.line(&format!("Num {}", n.canon())),
            Expr::Ident(id) => self.line(&format!("Ident {}", id.name)),
            Expr::All(_) => self.line("All"),
            Expr::NoneKw(_) => self.line("None"),
            Expr::True(_) => self.line("True"),
            Expr::False(_) => self.line("False"),
            Expr::Unary { op, expr, .. } => {
                let name = match op {
                    UnaryOp::Neg => "-",
                    UnaryOp::Not => "not",
                };
                self.nested(&format!("Unary op={name}"), |d| d.expr(expr));
            }
            Expr::Binary { op, lhs, rhs, .. } => {
                self.nested(&format!("Binary op={}", op.as_str()), |d| {
                    d.expr(lhs);
                    d.expr(rhs);
                });
            }
            Expr::Cmp { op, lhs, rhs, .. } => {
                self.nested(&format!("Cmp op={}", op.as_str()), |d| {
                    d.expr(lhs);
                    d.expr(rhs);
                });
            }
            Expr::Band {
                kind, expr, lo, hi, ..
            } => {
                let k = match kind {
                    BandKind::In => "in",
                    BandKind::Out => "out",
                };
                self.nested(
                    &format!("Band kind={k} lo={} hi={}", lo.canon(), hi.canon()),
                    |d| d.expr(expr),
                );
            }
            Expr::Ternary {
                guard, then, els, ..
            } => {
                let header = format!("Ternary has_else={}", els.is_some());
                self.nested(&header, |d| {
                    d.expr(guard);
                    d.expr(then);
                    if let Some(els) = els {
                        d.expr(els);
                    }
                });
            }
            Expr::Call { name, args, .. } => {
                self.nested(&format!("Call name={}", name.name), |d| {
                    for arg in args {
                        d.arg(arg);
                    }
                });
            }
            Expr::Dot { base, field, .. } => {
                self.nested(&format!("Dot field={}", field.name), |d| d.expr(base));
            }
            Expr::Index { base, index, .. } => {
                self.nested(&format!("Index {}", index.canon()), |d| d.expr(base));
            }
            Expr::Slice {
                base, start, end, ..
            } => {
                let s = start.map(|i| i.canon()).unwrap_or_default();
                let e = end.map(|i| i.canon()).unwrap_or_default();
                self.nested(&format!("Slice {s}:{e}"), |d| d.expr(base));
            }
            Expr::UnderscoreIndex { base, index, .. } => {
                self.nested(&format!("UIndex {}", index.canon()), |d| d.expr(base));
            }
            Expr::UnderscoreAll { base, .. } => {
                self.nested("UAll", |d| d.expr(base));
            }
            Expr::Abs { expr, .. } => {
                self.nested("Abs", |d| d.expr(expr));
            }
            Expr::Braced { args, prop, .. } => {
                self.nested(&format!("Braced prop={}", prop.name), |d| {
                    for arg in args {
                        d.arg(arg);
                    }
                });
            }
            Expr::ParticleList { items, .. } => {
                self.nested("ParticleList", |d| {
                    for item in items {
                        d.expr(item);
                    }
                });
            }
            Expr::Error(_) => self.line("Error"),
        }
    }
}

fn info_item(item: &InfoItem) -> String {
    match item {
        InfoItem::Ident(id) => format!("id:{}", id.name),
        InfoItem::Str(s) => format!("str:{:?}", s.value),
        InfoItem::Num(n) => format!("num:{}", n.canon()),
    }
}

File: reimplementation/smash_cpp2/libs/syntax/src/dump.cpp
#include "adl2/syntax/dump.hpp"

#include <cstdio>
#include <functional>
#include <sstream>

namespace adl2::syntax {

std::string rust_debug_str(std::string_view s) {
  // Rust `{:?}` for `str` (`char::escape_debug`): `\t \n \r \0 \\ \"` get a
  // backslash form, `'` stays as-is, and other control characters (C0, DEL,
  // C1 = U+0080..U+009F) print as `\u{x}` with unpadded lowercase hex. Other
  // non-printable Unicode (soft hyphen, zero-width, combining marks) is
  // passed through; the corpus has none and the tables are not worth it.
  std::string out;
  out.reserve(s.size() + 2);
  out.push_back('"');
  auto unicode_escape = [&](unsigned int cp) {
    char buf[16];
    std::snprintf(buf, sizeof buf, "\\u{%x}", cp);
    out += buf;
  };
  for (std::size_t i = 0; i < s.size(); ++i) {
    const auto c = static_cast<unsigned char>(s[i]);
    switch (c) {
      case '\t': out += "\\t"; continue;
      case '\n': out += "\\n"; continue;
      case '\r': out += "\\r"; continue;
      case '\0': out += "\\0"; continue;
      case '\\': out += "\\\\"; continue;
      case '"': out += "\\\""; continue;
      default: break;
    }
    if (c < 0x20 || c == 0x7F) {
      unicode_escape(c);
      continue;
    }
    if (c == 0xC2 && i + 1 < s.size()) {
      const auto d = static_cast<unsigned char>(s[i + 1]);
      if (d >= 0x80 && d <= 0x9F) {
        unicode_escape(d);
        ++i;
        continue;
      }
    }
    out.push_back(static_cast<char>(c));
  }
  out.push_back('"');
  return out;
}

namespace {

struct Dumper {
  std::string out;
  int depth = 0;

  void line(const std::string& text) {
    for (int i = 0; i < depth; ++i) out += "  ";
    out += text;
    out.push_back('\n');
  }

  static std::string at(const Span& sp) {
    return "@" + std::to_string(sp.line) + ":" + std::to_string(sp.column);
  }

  void nested(const std::string& header, const std::function<void()>& f) {
    line(header);
    ++depth;
    f();
    --depth;
  }

  void dump_arg(const Arg& a) {
    switch (a.kind) {
      case Arg::Kind::Expr:
        if (a.expr) dump_expr(*a.expr);
        break;
      case Arg::Kind::Str:
        line("Str " + rust_debug_str(a.str.value));
        break;
      case Arg::Kind::Path:
        line("Path " + rust_debug_str(a.str.value));
        break;
    }
  }

  void dump_expr(const Expr& e) {
    switch (e.kind) {
      case ExprKind::Num:
        line("Num " + e.num.canon());
        break;
      case ExprKind::Ident:
        line("Ident " + e.ident.name);
        break;
      case ExprKind::All:
        line("All");
        break;
      case ExprKind::NoneKw:
        line("None");
        break;
      case ExprKind::True:
        line("True");
        break;
      case ExprKind::False:
        line("False");
        break;
      case ExprKind::Unary: {
        const char* name = e.unary_op == UnaryOp::Neg ? "-" : "not";
        nested(std::string("Unary op=") + name, [&] {
          if (e.child) dump_expr(*e.child);
        });
        break;
      }
      case ExprKind::Binary:
        nested(std::string("Binary op=") + bin_op_str(e.bin_op), [&] {
          if (e.lhs) dump_expr(*e.lhs);
          if (e.rhs) dump_expr(*e.rhs);
        });
        break;
      case ExprKind::Cmp:
        nested(std::string("Cmp op=") + cmp_op_str(e.cmp_op), [&] {
          if (e.lhs) dump_expr(*e.lhs);
          if (e.rhs) dump_expr(*e.rhs);
        });
        break;
      case ExprKind::Band: {
        const char* k = e.band_kind == BandKind::In ? "in" : "out";
        nested("Band kind=" + std::string(k) + " lo=" + e.band_lo.canon() +
                   " hi=" + e.band_hi.canon(),
               [&] {
                 if (e.child) dump_expr(*e.child);
               });
        break;
      }
      case ExprKind::Ternary:
        nested("Ternary has_else=" +
                   std::string(e.ternary_has_else ? "true" : "false"),
               [&] {
                 if (e.guard) dump_expr(*e.guard);
                 if (e.then_e) dump_expr(*e.then_e);
                 if (e.else_e) dump_expr(*e.else_e);
               });
        break;
      case ExprKind::Call:
        nested("Call name=" + e.field.name, [&] {
          for (const auto& a : e.args) dump_arg(*a);
        });
        break;
      case ExprKind::Dot:
        nested("Dot field=" + e.field.name, [&] {
          if (e.child) dump_expr(*e.child);
        });
        break;
      case ExprKind::Member:
        nested("Member field=" + e.field.name, [&] {
          if (e.child) dump_expr(*e.child);
        });
        break;
      case ExprKind::Index:
        nested("Index " + e.index.canon(), [&] {
          if (e.child) dump_expr(*e.child);
        });
        break;
      case ExprKind::Slice: {
        std::string s =
            e.slice_start ? e.slice_start->canon() : std::string();
        std::string end =
            e.slice_end ? e.slice_end->canon() : std::string();
        nested("Slice " + s + ":" + end, [&] {
          if (e.child) dump_expr(*e.child);
        });
        break;
      }
      case ExprKind::UnderscoreIndex:
        nested("UIndex " + e.index.canon(), [&] {
          if (e.child) dump_expr(*e.child);
        });
        break;
      case ExprKind::UnderscoreAll:
        nested("UAll", [&] {
          if (e.child) dump_expr(*e.child);
        });
        break;
      case ExprKind::Abs:
        nested("Abs", [&] {
          if (e.child) dump_expr(*e.child);
        });
        break;
      case ExprKind::Braced:
        nested("Braced prop=" + e.field.name, [&] {
          for (const auto& a : e.args) dump_arg(*a);
        });
        break;
      case ExprKind::ParticleList:
        nested("ParticleList", [&] {
          for (const auto& it : e.items) dump_expr(*it);
        });
        break;
      case ExprKind::Error:
        line("Error");
        break;
    }
  }

  void dump_object_stmt(const ObjectStmt& stmt) {
    switch (stmt.kind) {
      case ObjectStmt::Kind::Take: {
        std::string header = "Take kw=" + stmt.keyword;
        switch (stmt.take_source.kind) {
          case TakeSourceKind::Ident:
            header += " src=" + stmt.take_source.name.name;
            break;
          case TakeSourceKind::Call:
            header += " src=call:" + stmt.take_source.name.name;
            break;
          case TakeSourceKind::Union: {
            header += " src=union(";
            for (std::size_t i = 0; i < stmt.take_source.members.size(); ++i) {
              if (i) header += ",";
              header += stmt.take_source.members[i].name;
            }
            header += ")";
            break;
          }
          case TakeSourceKind::Expr:
            header += " src=expr";
            break;
        }
        if (!stmt.binders.empty()) {
          header += " binders=[";
          for (std::size_t i = 0; i < stmt.binders.size(); ++i) {
            if (i) header += ",";
            header += stmt.binders[i].name;
          }
          header += "]";
        }
        if (stmt.alias) header += " alias=" + stmt.alias->name;
        header += " " + at(stmt.span);
        if (stmt.take_source.kind == TakeSourceKind::Call) {
          nested(header, [&] {
            for (const auto& a : stmt.take_source.args) dump_arg(*a);
          });
        } else {
          line(header);
        }
        break;
      }
      case ObjectStmt::Kind::Cut:
        nested("Cut kw=" + stmt.keyword + " " + at(stmt.span), [&] {
          if (stmt.cond) dump_expr(*stmt.cond);
        });
        break;
      case ObjectStmt::Kind::Reject:
        nested("Reject " + at(stmt.span), [&] {
          if (stmt.cond) dump_expr(*stmt.cond);
        });
        break;
      case ObjectStmt::Kind::Derived:
        nested("Derived kw=" + stmt.keyword + " name=" + stmt.name.name + " " +
                   at(stmt.span),
               [&] {
                 if (stmt.body) dump_expr(*stmt.body);
               });
        break;
      case ObjectStmt::Kind::Define:
        nested("Define kw=" + stmt.define.keyword + " name=" +
                   stmt.define.name.name + " " + at(stmt.define.span),
               [&] {
                 if (stmt.define.body) dump_expr(*stmt.define.body);
               });
        break;
    }
  }

  void dump_region_stmt(const RegionStmt& stmt) {
    switch (stmt.kind) {
      case RegionStmt::Kind::Cut:
        nested("Cut kw=" + stmt.keyword + " " + at(stmt.span), [&] {
          if (stmt.cond) dump_expr(*stmt.cond);
        });
        break;
      case RegionStmt::Kind::Reject:
        nested("Reject " + at(stmt.span), [&] {
          if (stmt.cond) dump_expr(*stmt.cond);
        });
        break;
      case RegionStmt::Kind::RegionRef:
        line("RegionRef name=" + stmt.name.name + " " + at(stmt.span));
        break;
      case RegionStmt::Kind::Bin: {
        std::string header = "Bin";
        if (stmt.label) header += " label=" + rust_debug_str(stmt.label->value);
        header += " " + at(stmt.span);
        nested(header, [&] {
          if (stmt.bin_body.kind == BinBodyKind::Boundaries) {
            std::string es;
            for (std::size_t i = 0; i < stmt.bin_body.edges.size(); ++i) {
              if (i) es += " ";
              es += stmt.bin_body.edges[i].canon();
            }
            nested("Boundaries edges=[" + es + "]", [&] {
              if (stmt.bin_body.var) dump_expr(*stmt.bin_body.var);
            });
          } else if (stmt.bin_body.cond) {
            dump_expr(*stmt.bin_body.cond);
          }
        });
        break;
      }
      case RegionStmt::Kind::Trigger:
        nested("Trigger " + at(stmt.span), [&] {
          if (stmt.cond) dump_expr(*stmt.cond);
        });
        break;
      case RegionStmt::Kind::Histo:
        nested("Histo name=" + stmt.name.name + " title=" +
                   rust_debug_str(stmt.title.value) + " " + at(stmt.span),
               [&] {
                 for (const auto& a : stmt.histo_args) {
                   switch (a.kind) {
                     case HistoArgKind::Num:
                       line("Num " + a.num.canon());
                       break;
                     case HistoArgKind::NumList: {
                       std::string es;
                       for (std::size_t i = 0; i < a.nums.size(); ++i) {
                         if (i) es += " ";
                         es += a.nums[i].canon();
                       }
                       line("NumList [" + es + "]");
                       break;
                     }
                     case HistoArgKind::Expr:
                       if (a.expr) dump_expr(*a.expr);
                       break;
                   }
                 }
               });
        break;
      case RegionStmt::Kind::Weight:
        nested("Weight name=" + stmt.name.name + " " + at(stmt.span), [&] {
          if (stmt.weight_value.kind == WeightValueKind::Num) {
            line("Num " + stmt.weight_value.num.canon());
          } else if (stmt.weight_value.expr) {
            dump_expr(*stmt.weight_value.expr);
          }
        });
        break;
      case RegionStmt::Kind::Save:
        nested("Save name=" + stmt.name.name + " format=" + stmt.format.name +
                   " " + at(stmt.span),
               [&] {
                 for (const auto& a : stmt.args) dump_arg(*a);
               });
        break;
      case RegionStmt::Kind::Print:
        nested("Print " + at(stmt.span), [&] {
          for (const auto& a : stmt.args) dump_arg(*a);
        });
        break;
      case RegionStmt::Kind::Counts: {
        std::string items;
        for (std::size_t i = 0; i < stmt.counts_items.size(); ++i) {
          if (i) items += " ";
          items += stmt.counts_items[i];
        }
        line("Counts format=" + stmt.format.name + " items=[" + items + "] " +
             at(stmt.span));
        break;
      }
      case RegionStmt::Kind::Sort:
        line("Sort (unsupported) raw=" + rust_debug_str(stmt.sort_raw) + " " +
             at(stmt.span));
        break;
      case RegionStmt::Kind::TypeTag:
        line("TypeTag value=" + stmt.type_value.name + " " + at(stmt.span));
        break;
    }
  }

  void dump_section(const Section& s) {
    switch (s.kind) {
      case SectionKind::Info:
        nested("Info name=" + s.info.name.name + " " + at(s.info.span), [&] {
          for (const auto& line_ : s.info.lines) {
            line("Line key=" + line_.key.name +
                 " value=" + rust_debug_str(line_.value));
          }
        });
        break;
      case SectionKind::Table: {
        const auto& t = s.table;
        nested("Table name=" + t.name.name + " type=" + t.table_type.name +
                   " nvars=" + std::to_string(t.nvars) +
                   " errors=" + std::string(t.errors ? "true" : "false") + " " +
                   at(t.span),
               [&] {
                 std::string vals;
                 for (std::size_t i = 0; i < t.values.size(); ++i) {
                   if (i) vals += " ";
                   vals += t.values[i].canon();
                 }
                 line("Values [" + vals + "]");
               });
        break;
      }
      case SectionKind::CountsFormat:
        nested("CountsFormat name=" + s.counts_format.name.name + " " +
                   at(s.counts_format.span),
               [&] {
                 for (const auto& p : s.counts_format.processes) {
                   std::string cols;
                   for (std::size_t i = 0; i < p.columns.size(); ++i) {
                     if (i) cols += ", ";
                     cols += p.columns[i].name;
                   }
                   line("Process name=" + p.name.name + " title=" +
                        rust_debug_str(p.title.value) + " columns=[" + cols +
                        "]");
                 }
               });
        break;
      case SectionKind::Define:
        nested("Define kw=" + s.define.keyword + " name=" + s.define.name.name +
                   " " + at(s.define.span),
               [&] {
                 if (s.define.body) dump_expr(*s.define.body);
               });
        break;
      case SectionKind::Object:
        nested("Object kw=" + std::string(object_kw_str(s.object.keyword)) +
                   " name=" + s.object.name.name + " " + at(s.object.span),
               [&] {
                 for (const auto& st : s.object.stmts) dump_object_stmt(st);
               });
        break;
      case SectionKind::Region:
        nested("Region kw=" + std::string(region_kw_str(s.region.keyword)) +
                   " name=" + s.region.name.name + " " + at(s.region.span),
               [&] {
                 for (const auto& st : s.region.stmts) dump_region_stmt(st);
               });
        break;
    }
  }
};

}  // namespace

std::string dump_ast(std::string_view /*src*/, const FileAst& file) {
  Dumper d;
  d.line("File");
  ++d.depth;
  for (const auto& section : file.sections) {
    d.dump_section(section);
  }
  return d.out;
}

}  // namespace adl2::syntax
